paddle/fluid/operators/ngraph/ngraph_engine.cc

```diff
@@ -485,99 +485,105 @@ void NgraphEngine::BuildNgFunction(const framework::ExecutionContext& ctx) {
     func_inputs.emplace_back(prm);
   }
 
-  ngraph_function_ =
-      std::make_shared<ngraph::Function>(func_outputs, func_inputs);
+  return std::make_shared<ngraph::Function>(func_outputs, func_inputs);
+}
+
+void NgraphEngine::ClearNgCache() {
+  auto it = engine_cache.begin();
+  while (it != engine_cache.end()) {
+    auto ng_engine = it->second;
+    backend_->remove_compiled_function(ng_engine.ngraph_handle);
+    ++it;
+  }
+  engine_cache.clear();
+  auto it_tensor = t_in_cache_.begin();
+  while (it_tensor != t_in_cache_.end()) {
+    auto t_vec = it_tensor->second;
+    for (auto t_in : t_vec) {
+      t_in.reset();
+    }
+    ++it_tensor;
+  }
+  t_in_cache_.clear();
 }
 
 void NgraphEngine::GetNgFunction(const framework::ExecutionContext& ctx) {
   auto interval = ctx.Attr<std::vector<int>>("interval");
   std::string engine_key = ctx.Attr<std::string>("engine_key");
+
+  // set to flase, to debug cache or recompile everytime.
   bool use_cache = true;
-  if (use_cache) {
-    this->func_cache_key_ = "";
-    for (int i = 0; i < static_cast<int>(feed_vars.size()); ++i) {
-      auto* var = scope_.FindVar(feed_vars[i]);
-      if (var && var->IsType<framework::LoDTensor>()) {
-        auto* tensor_pd = GetLoDTensorOrSelectedRowsValueFromVar(*var);
-        auto dims = tensor_pd->dims();
-        for (int j = 0; j < dims.size(); ++j) {
-          func_cache_key_ += std::to_string(dims[j]);
-        }
+  if (!use_cache) ClearNgCache();
+
+  this->func_cache_key_ = "";
+  for (int i = 0; i < static_cast<int>(feed_vars.size()); ++i) {
+    auto* var = scope_.FindVar(feed_vars[i]);
+    if (var && var->IsType<framework::LoDTensor>()) {
+      auto* tensor_pd = GetLoDTensorOrSelectedRowsValueFromVar(*var);
+      auto dims = tensor_pd->dims();
+      for (int j = 0; j < dims.size(); ++j) {
+        func_cache_key_ += std::to_string(dims[j]);
```

Comment:
For single instance case, it is enough, but for multi-instance case, this key generation condition may not be sufficient .

Reply:
we will update per multi-instance need when we get there